plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java

```diff
@@ -312,20 +313,20 @@ private List<DeltaLakeSplit> splitsForFile(
 
     private static String buildSplitPath(String tableLocation, AddFileEntry addAction)
     {
-        // paths are relative to the table location and are RFC 2396 URIs
+        // paths are relative to the table location or absolute in case of shallow cloned table and are RFC 2396 URIs
         // https://github.com/delta-io/delta/blob/master/PROTOCOL.md#add-file-and-remove-file
         URI uri = URI.create(addAction.getPath());
-        String path = uri.getPath();
 
+        if (uri.isAbsolute()) {
+            return uri.getScheme() + ":" + uri.getSchemeSpecificPart();
```

Comment:
- should we use `getRawSchemeSpecificPart` so that the returned String can be parsed back as URI?
- this chops off `#fragment` from original `addAction.getPath()`. Does it remove anything else?
  in fact, why not return `addAction.getPath()` here?

cc @electrum 

- we have abfs-specific code handling for relative paths. Why we don't need it for absolute paths? 
  (relates to 77cf3a437b486d87c714121d4822abf7daace971 / https://github.com/trinodb/trino/pull/15381)

cc @ebyhr

Reply:
>should we use getRawSchemeSpecificPart so that the returned String can be parsed back as URI?

yes @findepi, that's the intention

 >why not return addAction.getPath() here?

in case of special characters presence in the table basePath, if just `addAction.getPath()` is returned, then it gets encoded again only to throw `404/PathNotFoundException`

```
* actual table base path -  'abfs://ct@st.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%25/base_table_test_vin+/'
* encoded path in delta log  - 'abfs://ct@stdfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/base_table_test_vin+/part-00000-eeb717bb-9056-4959-bc79-a642879d8977-c000.snappy.parquet'

Exception encountered

Query 20230510_052556_00004_mttb2 failed: Error opening Hive split abfs://ct@st.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/base_table_test_vin+/part-00001-1aa100b2-0fa5-473d-abca-383af2e73b85-c000.snappy.parquet (offset=0, length=654): HEAD https://st.dfs.core.windows.net/ct/prod-data/dev_30.db/%2Bab/a%252525/base_table_test_vin%2B/part-00001-1aa100b2-0fa5-473d-abca-383af2e73b85-c000.snappy.parquet?timeout=90
StatusCode=404
StatusDescription=The specified path does not exist.
ErrorCode=
ErrorMessage=
io.trino.spi.TrinoException: Error opening Hive split abfs://ct@st.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/base_table_test_vin+/part-00001-1aa100b2-0fa5-473d-abca-383af2e73b85-c000.snappy.parquet (offset=0, length=654): HEAD https://st.dfs.core.windows.net/ct/prod-data/dev_30.db/%2Bab/a%252525/base_table_test_vin%2B/part-00001-1aa100b2-0fa5-473d-abca-383af2e73b85-c000.snappy.parquet?timeout=90
StatusCode=404
StatusDescription=The specified path does not exist.
ErrorCode=
ErrorMessage=
	at io.trino.plugin.hive.parquet.ParquetPageSourceFactory.createPageSource(ParquetPageSourceFactory.java:313)
	at io.trino.plugin.deltalake.DeltaLakePageSourceProvider.createPageSource(DeltaLakePageSourceProvider.java:203)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:49)
	at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:62)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:298)
	at io.trino.operator.Driver.processInternal(Driver.java:402)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:305)
	at io.trino.operator.Driver.tryWithLock(Driver.java:701)
	at io.trino.operator.Driver.process(Driver.java:297)
	at io.trino.operator.Driver.processForDuration(Driver.java:268)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:888)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:187)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:561)
	at io.trino.$gen.Trino_dev____20230510_052255_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.io.FileNotFoundException: HEAD https://st.dfs.core.windows.net/ct/prod-data/dev_30.db/%2Bab/a%252525/base_table_test_vin%2B/part-00001-1aa100b2-0fa5-473d-abca-383af2e73b85-c000.snappy.parquet?timeout=90
StatusCode=404
StatusDescription=The specified path does not exist.
ErrorCode=
ErrorMessage=
	at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem.checkException(AzureBlobFileSystem.java:926)
	at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem.open(AzureBlobFileSystem.java:177)
	at io.trino.hdfs.TrinoFileSystemCache$FileSystemWrapper.open(TrinoFileSystemCache.java:340)
	at org.apache.hadoop.fs.FileSystem.open(FileSystem.java:906)
	at io.trino.filesystem.hdfs.HdfsInputFile.lambda$openFile$1(HdfsInputFile.java:108)
	at io.trino.hdfs.authentication.NoHdfsAuthentication.doAs(NoHdfsAuthentication.java:25)
	at io.trino.hdfs.HdfsEnvironment.doAs(HdfsEnvironment.java:126)
	at io.trino.filesystem.hdfs.HdfsInputFile.openFile(HdfsInputFile.java:108)
	at io.trino.filesystem.hdfs.HdfsInputFile.newInput(HdfsInputFile.java:57)
	at io.trino.filesystem.tracing.TracingInputFile.lambda$newInput$0(TracingInputFile.java:52)
	at io.trino.filesystem.tracing.Tracing.withTracing(Tracing.java:47)
	at io.trino.filesystem.tracing.TracingInputFile.newInput(TracingInputFile.java:52)
	at io.trino.plugin.hive.parquet.TrinoParquetDataSource.<init>(TrinoParquetDataSource.java:39)
	at io.trino.plugin.hive.parquet.ParquetPageSourceFactory.createPageSource(ParquetPageSourceFactory.java:228)
	... 16 more
Caused by: org.apache.hadoop.fs.azurebfs.contracts.exceptions.AbfsRestOperationException: HEAD https://st.dfs.core.windows.net/ct/prod-data/dev_30.db/%2Bab/a%252525/base_table_test_vin%2B/part-00001-1aa100b2-0fa5-473d-abca-383af2e73b85-c000.snappy.parquet?timeout=90
StatusCode=404
StatusDescription=The specified path does not exist.
ErrorCode=
ErrorMessage=
	at org.apache.hadoop.fs.azurebfs.services.AbfsRestOperation.execute(AbfsRestOperation.java:134)
	at org.apache.hadoop.fs.azurebfs.services.AbfsClient.getPathProperties(AbfsClient.java:352)
	at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystemStore.openFileForRead(AzureBlobFileSystemStore.java:349)
	at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem.open(AzureBlobFileSystem.java:174)
	... 28 more
```

Reply:
>this chops off #fragment from original addAction.getPath(). Does it remove anything else?

@findepi indeed `#getRawSchemeSpecificPart()` chops off `#fragment`, we can handle that with `+ "#" + uri.getFragment()`, where as the current impl `#getPath()` doesn't contain `#query as well as #fragment`

Reply:
> `#getRawSchemeSpecificPart()` chops off `#fragment` only, we can handle that with `uri.getScheme() + ":" + uri.getSchemeSpecificPart() + "#" + uri.getFragment()`

thanks for confirming that

what was the reason not return `addAction.getPath()` here?

> > why not return addAction.getPath() here?
> 
> in case of special characters presence in the table basePath, if just `addAction.getPath()` is returned, then it gets encoded again

where does "again" come from?

Reply:
>where does "again" come from?

@findepi it's within [ABFS client code](https://github.com/apache/hadoop/blob/trunk/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java#L1295), it's again encoded, `% with %25` in this particular case

```
23/05/10 17:34:36 DEBUG services.AbfsClient: HttpRequest: GetAcl: 200,,,cid=1238024a-33f3-4a36-aab0-ab8075b7282d,rid=bf703d73-001f-006a-0f65-83bb39000000,sent=0,recv=0,HEAD,https://st.dfs.core.windows.net/ct/?upn=true&action=getAccessControl&timeout=90
23/05/10 17:34:36 DEBUG azurebfs.AzureBlobFileSystemStore: getFileStatus filesystem: ct path: abfs://ct@st.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/base_table_test_vin+/part-00000-eeb717bb-9056-4959-bc79-a642879d8977-c000.snappy.parquet isNamespaceEnabled: true
23/05/10 17:34:36 DEBUG services.AbfsClient: First execution of REST operation - GetPathStatus
23/05/10 17:34:36 DEBUG services.AbfsClient: HttpRequest: GetPathStatus: 404,,,cid=c9f579ba-0eb1-43f8-9e5d-8c89a95b1ebe,rid=bf703d76-001f-006a-1265-83bb39000000,sent=0,recv=0,HEAD,https://st.dfs.core.windows.net/ct/prod-data/dev_30.db/%2Bab/a%252525/base_table_test_vin%2B/part-00000-eeb717bb-9056-4959-bc79-a642879d8977-c000.snappy.parquet?upn=true&action=getStatus&timeout=90
ls: `abfs://ct@st.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/base_table_test_vin+/part-00000-eeb717bb-9056-4959-bc79-a642879d8977-c000.snappy.parquet': No such file or directory
23/05/10 17:34:36 DEBUG azurebfs.AzureBlobFileSystem: AzureBlobFileSystem.close
23/05/10 17:34:36 DEBUG azurebfs.AzureBlobFileSystem: Closing Abfs: AzureBlobFileSystem{uri=abfs://ct@st.dfs.core.windows.net, user='sshuser', primaryUserGroup='sshuser', Statistics: {{Context=AbfsContext}{AbfsID=0adeffa5-dc6a-4dd2-8c1f-188ae78c3d09}{AbfsBucket=st.dfs.core.windows.net}{op_create=0}{op_open=0}{op_get_file_status=1}{op_append=0}{op_create_non_recursive=0}{op_delete=0}{op_exists=0}{op_get_delegation_token=0}{op_list_status=0}{op_mkdirs=0}{op_rename=0}{directories_created=0}{directories_deleted=0}{files_created=0}{files_deleted=0}{error_ignored=0}{connections_made=2}{send_requests=0}{get_responses=2}{bytes_sent=0}{bytes_received=0}{read_throttles=0}{write_throttles=0}}}
```

Reply:
@findepi @ebyhr can `#fragments` be part of the FilePath? If yes then i'll change the code to handle the same.

Reply:
There are no fragments allowed in the Trino paths atm.

https://github.com/trinodb/trino/blob/58f9a527fc48859f2d74018a90529fd77711ee01/lib/trino-filesystem/src/main/java/io/trino/filesystem/Location.java#L65